sound/soc/sof/ipc4-pcm.c

```diff
@@ -32,65 +60,184 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
 }
 EXPORT_SYMBOL(sof_ipc4_set_pipeline_state);
 
+/* helper function to get the list of pipeline widgets that need to be triggered */
+static int
+sof_ipc4_get_pipelines_in_order(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget *widget,
```

Comment:
I don't understand how does it work since I am not good at recursion.  How about feedback stream like echo reference ?

As I know another method to sort pipelines is based on pipeline priority. Each pipeline has priority defined in topology and driver can sort pipeline with priority. But currently we don't set it in ipc4-topology.c

````c
	/* TODO: Get priority from topology */
	pipeline->priority = 0;
````

Reply:
priority has a different meaning that order @RanderWang. We may need to set the priority at some point later on to pass hints to the scheduler. We cannot mix up these 2 things

Reply:
I found windows driver use such priority to sort pipelines.